internal/fields/mappings.go

```diff
@@ -624,22 +576,61 @@ func (v *MappingValidator) validateMappingsNotInPreview(currentPath string, chil
 			continue
 		}
 
-		// TODO: validate mapping with dynamic templates first than validating with ECS
-		// just raise an error if both validation processes fail
+		// validate whether or not the field has a corresponding dynamic template
+		if len(rawDynamicTemplates) > 0 {
+			err := v.matchingWithDynamicTemplates(fieldPath, def, dynamicTemplates)
+			if err == nil {
+				continue
+			}
+		}
 
-		// are all fields under this key defined in ECS?
-		err = v.validateMappingInECSSchema(fieldPath, def)
-		if err != nil {
-			logger.Warnf("undefined path %q (pending to check dynamic templates)", fieldPath)
-			errs = append(errs, fmt.Errorf("field %q is undefined: %w", fieldPath, err))
+		// validate whether or not all fields under this key are defined in ECS
+		ecsErrs := v.validateMappingInECSSchema(fieldPath, def)
+		if len(ecsErrs) > 0 {
+			for _, e := range ecsErrs {
+				errs = append(errs, fmt.Errorf("field %q is undefined: %w", fieldPath, e))
+			}
 		}
 	}
 	return errs.Unique()
 }
 
+// matchingWithDynamicTemplates validates a given definition (currentPath) with a set of dynamic templates.
+// The dynamic templates parameters are based on https://www.elastic.co/guide/en/elasticsearch/reference/8.17/dynamic-templates.html
+func (v *MappingValidator) matchingWithDynamicTemplates(currentPath string, definition map[string]any, dynamicTemplates []dynamicTemplate) error {
+	for _, template := range dynamicTemplates {
+		matches, err := template.Matches(currentPath, definition)
+		if err != nil {
+			return fmt.Errorf("failed to validate %q with dynamic template %q: %w", currentPath, template.name, err)
+		}
+
+		if !matches {
+			// Look for another dynamic template
+			continue
+		}
+
+		logger.Warnf("> Found dynamic template matched: %s", template.name)
+
+		logger.Debugf("> Check parameters (%q): %q", template.name, currentPath)
+		// Check that all parameters match (setting no dynamic templates to avoid recursion)
+		errs := v.validateObjectMappingAndParameters(template.mapping, definition, currentPath, []map[string]any{}, true)
+		if errs != nil {
+			// Look for another dynamic template
+			logger.Warnf(">> Invalid dynamic template for %q:\n%s", currentPath, errs.Unique())
+			continue
+		}
+
+		logger.Warnf("> Valid template for path %q: %q", currentPath, template.name)
```

Comment:
This is printed for any field that matches with a dynamic template, right? Can it be too verbose?

Reply:
Yes, this is too verbose. I'll remove this one along with other debug messages.